### ISSUE TITLE
Regression with circe 0.6.1

### ISSUE THREAD

**Issue by jsancio** [circe/circe#549]
In Circe 0.5.4 the following code doesn't fail:

```scala
import cats.data.Xor
import io.circe.Decoder                                                                                               
import io.circe.DecodingFailure                                                                                       
import io.circe.generic.semiauto._                                                                                    
import io.circe.jawn.decode                                                                                           
                                                                                                                      
case class B(s: String)                                                                                               
object B {                                                                                                            
  val error = DecodingFailure("propagate this", Nil)                                                                  
  implicit val decoder: Decoder[B] = Decoder.failed[B](error)                                                         
}                                                                                                                     
                                                                                                                      
case class A(b: List[B])                                                                                              
object A {                                                                                                            
  implicit val decoder: Decoder[A] = deriveDecoder[A]                                                                 
}                                                                                                                                                                                                                                           
                                                                                                                      
object Test {                                                                                                         
  def main(args: Array[String]): Unit = {                                                                             
    val json = """                                                                                                    
    {                                                                                                                 
      "b": [ "" ]                                                                                                     
    }                                                                                                                 
    """                                                                                                               
                                                                                                                      
    val result = decode[A](json)                                                                                      
    assert(result == Xor.Left(B.error), s"result: $result")                            
  }                                                                                                                   
} 
```

While in Circe 0.6.1 the following code fails:

```scala
import io.circe.Decoder                                                                                               
import io.circe.DecodingFailure                                                                                       
import io.circe.generic.semiauto._                                                                                    
import io.circe.jawn.decode                                                                                           
                                                                                                                      
case class B(s: String)                                                                                               
object B {                                                                                                            
  val error = DecodingFailure("propagate this", Nil)                                                                  
  implicit val decoder: Decoder[B] = Decoder.failed[B](error)                                                         
}                                                                                                                     
                                                                                                                      
case class A(b: List[B])                                                                                              
object A {                                                                                                            
  implicit val decoder: Decoder[A] = deriveDecoder[A]                                                                 
}                                                                                                                                                                                                                                           
                                                                                                                      
object Test {                                                                                                         
  def main(args: Array[String]): Unit = {                                                                             
    val json = """                                                                                                    
    {                                                                                                                 
      "b": [ "" ]                                                                                                     
    }                                                                                                                 
    """                                                                                                               
                                                                                                                      
    val result = decode[A](json)                                                                                      
    assert(result == Left(B.error), s"result: $result")                                                               
  }                                                                                                                   
}
```

This is what the circe 0.6.1 code returns:

```
[error] (run-main-d) java.lang.AssertionError: assertion failed: result: Left(DecodingFailure([A]List[A], List()))
java.lang.AssertionError: assertion failed: result: Left(DecodingFailure([A]List[A], List()))
        at scala.Predef$.assert(Predef.scala:170)
        at Test$.main(Test.scala:27)
        at Test.main(Test.scala)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
```

**Comment by travisbrown** [comment]
I'm not sure I want to call this a regression, since the exact contents of the error message aren't documented behavior, but I understand that changes like this can be inconvenient.

The difference was introduced in #442, which provides `decodeList` instances for convenience of explicit use. I'd be happy to review / take into consideration a PR that removes the `withErrorMessage` for those instances.

In the meantime you can get the pre-0.6 error message by putting an instance like this into scope:

```scala
implicit def decodeList[A: Decoder]: Decoder[List[A]] = Decoder.decodeCanBuildFrom[A, List]
```

It's not ideal but it works.

**Comment by jsancio** [comment]
The workaround worked of us. Thanks!

**Comment by travisbrown** [comment]
Fixed by #649.
